packages/block-editor/src/components/inserter/block-patterns-tab.js

```diff
@@ -205,24 +310,60 @@ function BlockPatternsTabs( {
 	rootClientId,
 } ) {
 	const [ showPatternsExplorer, setShowPatternsExplorer ] = useState( false );
-	const categories = usePatternsCategories( rootClientId );
+	const [ patternSourceFilter, setPatternSourceFilter ] = useState( 'all' );
+	const patternSyncFilter = useSelect( ( select ) => {
+		const { getSettings } = select( blockEditorStore );
+		const settings = getSettings();
+		return settings.patternsSyncFilter;
+	}, [] );
+	const previousSyncFilter = usePrevious( patternSyncFilter );
+
+	// If the sync filter changes, we need to select the "All" category to avoid
+	// showing a confusing no results screen.
+	useEffect( () => {
+		if ( patternSyncFilter && patternSyncFilter !== previousSyncFilter ) {
+			onSelectCategory( allPatternsCategory, patternSourceFilter );
+		}
+	}, [
+		patternSyncFilter,
+		previousSyncFilter,
+		onSelectCategory,
+		patternSourceFilter,
+	] );
```

Comment:
Nit: Same here. Could we directly set the state in the render phase to avoid an extra render? Never mind if it's not possible though!